### PR TITLE
Method of character level evaluation for PII recognition

### PR DESCRIPTION
**Pull request by gabechu**
### Description
This PR shows one method in estimating a PII recognition system.

Basically, the method alternatively iterates through the predictions and ground truth for calculating precisions and recalls respectively for every entity being iterated.

**Risk**
May missed some of the corner cases, but we can only know about it when applying on real data.

#### Checklist
- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.
- [x] Rename `span_labels`

### CODE REVIEW

File: pii_recognition/evaluation/character_level_evaluation.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            )
+        label_name = span.entity_type
+        try:
+            label_code = label_to_int[label_name]
+        except KeyError as err:
+            raise Exception(
+                f"Label {str(err)} is not presented in 'label_to_int' mapping."
+            )
+
+        coding[s:e] = [label_code] * (e - s)
```
Comment (chausler) [inline pii_recognition/evaluation/character_level_evaluation.py:51]: is`span.end` included in the span?  ie. is the span assumed to be a closed interval? (https://mathworld.wolfram.com/Interval.html)

File: pii_recognition/evaluation/character_level_evaluation.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    compute_for: str,
+) -> List[Dict]:
+    if compute_for not in {"precision", "recall"}:
+        raise ValueError(
+            f"Available options for computation are: "
+            f"'precision', 'recall' but got {compute_for}"
+        )
+
+    config = {
+        "precision": {"loop": pred_spans, "no_loop": true_spans},
```
Comment (chausler) [inline pii_recognition/evaluation/character_level_evaluation.py:70]: what does `loop` and `no_loop` mean?

File: pii_recognition/evaluation/character_level_evaluation.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        text_length, config[compute_for]["no_loop"], label_mapping
+    )
+
+    scores: List = []
+    for span in config[compute_for]["loop"]:
+        span_coding: List[int] = encode_labels(text_length, [span], label_mapping)
+        int_label: int = label_mapping[span.entity_type]
+
+        if compute_for == "precision":
+            score = compute_label_precision(coding, span_coding, int_label)
```
Comment (chausler) [inline pii_recognition/evaluation/character_level_evaluation.py:84]: won't this compare the full length of the text instead of just the text for the span?

File: pii_recognition/evaluation/character_level_evaluation.py (hunk 2 above)
Comment (gabechu) [reply pii_recognition/evaluation/character_level_evaluation.py:70]: `loop` refers to the item we are going to take to the loop. This is a bit confused as there is item swapping and looping here, I would make changes to this replacing it with more readable and straightforward code.

**Comment by gabechu** [comment]
Just a note for me. Rethink of names, `span_labels` is not a clear and intuitive name, maybe use `entity` instead.

File: pii_recognition/evaluation/character_level_evaluation.py (hunk 1 above)
Comment (gabechu) [reply pii_recognition/evaluation/character_level_evaluation.py:51]: `end` is not included.

File: pii_recognition/evaluation/character_level_evaluation.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        elif compute_for == "recall":
+            score = compute_label_recall(span_coding, coding, int_label)
```
Comment (arwengriffioen) [inline pii_recognition/evaluation/character_level_evaluation.py:86]: compute_label_precision and compute_label_recall have different argument orderings? If this isn't a mistake could you make them the same order?

File: pii_recognition/evaluation/character_level_evaluation.py (hunk 3 above)
Comment (gabechu) [reply pii_recognition/evaluation/character_level_evaluation.py:84]: This will be replaced with the coming change, please refer to the above comment.

File: pii_recognition/evaluation/character_level_evaluation.py (hunk 4 above)
Comment (gabechu) [reply pii_recognition/evaluation/character_level_evaluation.py:86]: I had thought about that but didn't come up with a solution. `y_true` (1st arg of the function) varies depending on which score you'd like to calculate. Regarding to precision calculation, `y_true` is the coding derived from all true entities; regarding to recall calculation, `y_true` is the coding derived from from a particular predicted entity.

One option is to write out two loops that will increase readability, however, will have repetitive code.

Comment (arwengriffioen) [reply pii_recognition/evaluation/character_level_evaluation.py]: hrmm.  maybe rename the variables you are passing in? or use explicit arg assignment i.e. `y_true=coding`

**Review by arwengriffioen — APPROVED** [review]
lgtm. next time could you do the rename in a different PR? 21 files is a lot

**Comment by gabechu** [comment]
@arwengriffioen  @chausler Made a change that separates the calculation of precision and recall in which each of the two has a loop that iterate over a list of targeted entities and doing similar calculations. This change increases the readability but may look a bit repetitive, however, for evaluation metrics readability could be much important.

**Comment by gabechu** [comment]
> lgtm. next time could you do the rename in a different PR? 21 files is a lot

Absolutely. Indeed 21 files that's huge amount of changes.

File: pii_recognition/evaluation/character_level_evaluation.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    precisions = []
+    for pred_entity in pred_entities:
+        pred_entity_code: List[int] = label_encoder(
+            text_length, [pred_entity], label_mapping
+        )
+        int_label: int = label_mapping[pred_entity.entity_type]
+        score = compute_label_precision(true_code, pred_entity_code, int_label)
+
+        span_dict = asdict(pred_entity)
```
Comment (chausler) [inline pii_recognition/evaluation/character_level_evaluation.py:73]: i think we should call this `entity_dict` for naming consistency

File: pii_recognition/evaluation/character_level_evaluation.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+    recalls = []
+    for true_entity in true_entities:
+        true_entity_code: List[int] = label_encoder(
+            text_length, [true_entity], label_mapping
+        )
+        int_label: int = label_mapping[true_entity.entity_type]
+        score = compute_label_recall(true_entity_code, pred_code, int_label)
+
+        span_dict = asdict(true_entity)
```
Comment (chausler) [inline pii_recognition/evaluation/character_level_evaluation.py:97]: i think we should call this entity_dict for naming consistency

File: pii_recognition/evaluation/character_level_evaluation_test.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    spans = [Entity(entity_type="LOC", start=3, end=7)]
+
+    with pytest.raises(ValueError) as error:
+        label_encoder(5, spans, {"LOC": 1})
+    assert str(error.value) == (
+        "Span index is out of range: text length is 5 but got span index 7."
+    )
+
+
+def test_compute_precisions_recalls_for_exact_match():
```
Comment (chausler) [inline pii_recognition/evaluation/character_level_evaluation_test.py:56]: I think you should expand this test to include more than one entity and more than one entity type. This will better test the loop logic as well as make sure it works in a multiclass scope

**Review by chausler — APPROVED** [review]
on test and naming changes

**Comment by chausler** [comment]
ps. thanks for refactoring the code, it's much easier to follow now :-D

File: pii_recognition/evaluation/character_level_evaluation.py (hunk 6 above)
Comment (gabechu) [reply pii_recognition/evaluation/character_level_evaluation.py:97]: Good catch. Fixed.

File: pii_recognition/evaluation/character_level_evaluation.py (hunk 5 above)
Comment (gabechu) [reply pii_recognition/evaluation/character_level_evaluation.py:73]: Good catch. Fixed.

File: pii_recognition/evaluation/character_level_evaluation_test.py (hunk 7 above)
Comment (gabechu) [reply pii_recognition/evaluation/character_level_evaluation_test.py:54]: Done. All test cases updated regarding to precision and recall calculations.
